src/System.Net.HttpListener/tests/HttpListenerWebSocketTests.cs

```diff
@@ -0,0 +1,358 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Tests
+{
+    public class HttpListenerWebSocketTests : IDisposable
+    {
+        public static bool IsNotWindows7OrUapCore { get; } = !PlatformDetection.IsWindows7 && PlatformDetection.IsNotOneCoreUAP;
+        public static bool IsNotWindows7OrUapCoreAndIsWindowsImplementation { get; } = IsNotWindows7OrUapCore && Helpers.IsWindowsImplementation;
+
+        private HttpListenerFactory Factory { get; }
+        private HttpListener Listener { get; }
+        private ClientWebSocket Client { get; }
+        private Task ClientConnectTask { get; set; }
+
+        public HttpListenerWebSocketTests()
+        {
+            Factory = new HttpListenerFactory();
+            Listener = Factory.GetListener();
+            Client = new ClientWebSocket();
+        }
+
+        public void Dispose()
+        {
+            Factory.Dispose();
+            Client.Dispose();
+
+            if (ClientConnectTask?.IsCompleted == true)
+            {
+                ClientConnectTask?.Dispose();
+            }
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7OrUapCore))]
+        [InlineData(WebSocketMessageType.Text, false)]
+        [InlineData(WebSocketMessageType.Binary, false)]
+        [InlineData(WebSocketMessageType.Text, true)]
+        [InlineData(WebSocketMessageType.Binary, true)]
+        public async Task SendAsync_SendWholeBuffer_Success(WebSocketMessageType messageType, bool endOfMessage)
+        {
+            HttpListenerWebSocketContext context = await GetWebSocketContext();
+
+            await ClientConnectTask;
+            const string Text = "Hello Web Socket";
+            byte[] sentBytes = Encoding.ASCII.GetBytes(Text);
+
+            await context.WebSocket.SendAsync(new ArraySegment<byte>(sentBytes), messageType, endOfMessage, new CancellationToken());
+
+            byte[] receivedBytes = new byte[sentBytes.Length];
+            WebSocketReceiveResult result = await Client.ReceiveAsync(new ArraySegment<byte>(receivedBytes), new CancellationToken());
+            Assert.Equal(messageType, result.MessageType);
+            Assert.Equal(endOfMessage, result.EndOfMessage);
+            Assert.Null(result.CloseStatus);
+            Assert.Null(result.CloseStatusDescription);
+
+            Assert.Equal(Text, Encoding.ASCII.GetString(receivedBytes));
```

Comment:
There's technically no guarantee that you'll receive the same number of bytes in a single receive; it's valid for the implementation to split it up amongst several packets such that you'll get the data across multiple receives.  It's of course very likely this will pass, but it could be the source of sporadic failures in the future.